core/src/main/java/org/infinispan/protostream/impl/StringUtil.java

```diff
@@ -0,0 +1,155 @@
+package org.infinispan.protostream.impl;
+
+import static java.lang.invoke.MethodType.methodType;
+
+import java.lang.invoke.CallSite;
+import java.lang.invoke.LambdaMetafactory;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.function.Function;
+import java.util.function.ToIntFunction;
+
+import sun.misc.Unsafe;
+
+class StringUtil {
+
+   interface Utf8Helper {
+      boolean providesLatin1Bytes();
+      byte[] getBytes(String s);
+   }
+
+   private static final Log log = Log.LogFactory.getLog(ProtoStreamReaderImpl.class);
+   private static final Byte LATIN1 = (byte) 0;
+
+   static volatile Utf8Helper INSTANCE;
+
+   static Utf8Helper getUtf8Helper() {
+      if (INSTANCE == null) {
+         // Attempt to use IMPL_LOOKUP so that we utilise a Lambda - Most efficient
+         try {
+            INSTANCE = createLambdaHelper();
+            return INSTANCE;
+         } catch (Throwable t) {
+            log.warnf(t, "Unable to retrieve MethodHandles required for Lambda based String optimisations");
```

Comment:
Should these be WARN, is INFO not sufficient? Also I feel like we shouldn't log this if Unsafe isn't available as it will always be printed on startup and the user would have to downgrade to remove it.

Reply:
Also we need i18n for these messages.

Reply:
Updated so that we only output a WARN message if the `createFieldHelper` lookup fails.